spring-integration-core/src/main/java/org/springframework/integration/handler/BarrierMessageHandler.java

```diff
@@ -0,0 +1,192 @@
+	 */
+	public BarrierMessageHandler(long timeout) {
+		this(new HeaderAttributeCorrelationStrategy(IntegrationMessageHeaderAccessor.CORRELATION_ID),
+				new DefaultAggregatingMessageGroupProcessor(), timeout);
+	}
+
+	/**
+	 * Construct an instance with the provided timeout and output processor, and default
+	 * correlation strategy.
+	 * @param timeout the timeout in milliseconds.
+	 * @param outputProcessor the output {@link MessageGroupProcessor}.
+	 */
+	public BarrierMessageHandler(long timeout, MessageGroupProcessor outputProcessor) {
+		this(new HeaderAttributeCorrelationStrategy(IntegrationMessageHeaderAccessor.CORRELATION_ID),
+				outputProcessor, timeout);
+	}
+
+	/**
+	 * Construct an instance with the provided timeout and correlation strategy, and default
+	 * output processor.
+	 * @param timeout the timeout in milliseconds.
+	 * @param correlationStrategy the correlation strategy.
+	 */
+	public BarrierMessageHandler(CorrelationStrategy correlationStrategy, long timeout) {
+		this(correlationStrategy, new DefaultAggregatingMessageGroupProcessor(), timeout);
+	}
+
+	/**
+	 * Construct an instance with the provided timeout and output processor, and default
+	 * correlation strategy.
+	 * @param timeout the timeout in milliseconds.
+	 * @param correlationStrategy the correlation strategy.
+	 * @param outputProcessor the output {@link MessageGroupProcessor}.
+	 */
+	public BarrierMessageHandler(CorrelationStrategy correlationStrategy,
+			MessageGroupProcessor outputProcessor, long timeout) {
+		Assert.notNull(correlationStrategy, "'correlationStrategy' cannot be null");
+		Assert.notNull(outputProcessor, "'messageGroupProcessor' cannot be null");
+		this.correlationStrategy = correlationStrategy;
+		this.messageGroupProcessor =  outputProcessor;
+		this.timeout = timeout;
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Object key = this.correlationStrategy.getCorrelationKey(requestMessage);
+		if (key == null) {
+			throw new MessagingException(requestMessage, "Correlation Strategy returned null");
+		}
+		Thread existing = this.inProcess.putIfAbsent(key, Thread.currentThread());
+		if (existing != null) {
+			throw new MessagingException(requestMessage, "Correlation key ("
+					+ key + ") is already in use by " + existing.getName());
+		}
+		SynchronousQueue<Message<?>> syncQueue = createOrObtainQueue(key);
+		try {
+			Message<?> release = syncQueue.poll(this.timeout, TimeUnit.MILLISECONDS);
+			if (release != null) {
+				return processRelease(requestMessage, key, release);
+			}
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			throw new MessagingException(requestMessage, "Interrupted while waiting for release", e);
+		}
+		finally {
+			this.inProcess.remove(key);
+			this.suspensions.remove(key);
+		}
+		return null;
+	}
+
+	private Object processRelease(Message<?> requestMessage, Object key, Message<?> release) {
+		this.suspensions.remove(key);
```

Comment:
No; it's ok; we're protected by inProcess - but effectively, this one does not free up the key earlier because of `inProcess` so the conclusion is remove this one, I think.